networkit/cpp/io/NetworkitBinaryWriter.cpp

```diff
@@ -0,0 +1,201 @@
+
+uint64_t NetworkitBinaryWriter::splitNodes(uint64_t nodes, uint64_t chunks) {
+
+	nodesPerChunk = nodes / chunks;
+	return nodesPerChunk;
+}
+
+size_t NetworkitBinaryWriter::encode(uint64_t value, uint8_t* buffer) {
+	uint64_t n;
+	if(!value) {
+		buffer[0] =1;
+		return 1;
+	}
+
+	if (value > (uint64_t(1) << 63)) {
+		n = 8;
+		buffer[0] = 0;
+	} else {
+		n = (64 - tlx::clz(value)) / 7;
+		buffer[0] = (1 << n) | (value << (n + 1));
+		value >>= 8 - (n + 1);
+	}
+
+	for(int i = 0; i < n; i++) {
+		buffer[i+1] = value & 0xFF;
+	   	value >>= 8;
+	}
+	return n + 1;
+}
+
+uint64_t NetworkitBinaryWriter::encodeZigzag(int64_t value) {
+	uint64_t mapped = (value << 1) ^ (value >> 31);
+	return mapped;
+}
+
+void NetworkitBinaryWriter::write(const Graph &G, const std::string& path) {
+
+	std::ofstream outfile(path, std::ios::binary);
+	Aux::enforceOpened(outfile);
+
+	bool directed = G.isDirected();
+	uint64_t format = 0;
+	unsigned weight = 0;
+	Header header = {};
+
+	auto calculateBaseOffset = [&] () {
+		return sizeof(Header);
+	};
+
+	auto calculateAdjOffset = [&] () {
+		size_t off = nodes * sizeof(uint64_t); //prefSum
+		off += chunks * sizeof(uint64_t); //first vertex
+		return sizeof(Header) + off;
+	};
+
+	auto calculateWeightsOffset = [&] () {
+		size_t off = G.numberOfEdges() * sizeof(uint8_t); //adj arrays
+		off += chunks * sizeof(uint64_t);		  // adj offsets
+		return sizeof(Header) + off + calculateAdjOffset();
+	};
+
+	auto setFeatures = [&] () {
+		header.features |= (G.isDirected() & DIR_MASK);
+		header.features |= ((weight << WGHT_SHIFT) & WGHT_MASK);
+	};
+
+	auto setHeader = [&] () {
+		strncpy(header.magic,"nkbg000",8);
+		header.checksum = 0;
+		setFeatures();
+		header.nodes = nodes;
+		header.chunks = chunks;
+		header.offsetBaseData = calculateBaseOffset();
+		header.offsetAdjLists = calculateAdjOffset();
+		header.offsetWeights = calculateWeightsOffset();
+	};
+
+	auto writeHeader = [&] () {
+
+		outfile.write(header.magic, 8);
+		outfile.write(reinterpret_cast<char*>(&header.checksum), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.features), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.nodes), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.chunks), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetBaseData), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetAdjLists), sizeof(uint64_t));
+		outfile.write(reinterpret_cast<char*>(&header.offsetWeights), sizeof(uint64_t));
+	};
+
+	count prefSum = 0;
+	nodes = G.numberOfNodes();
+	if (nodes < chunks) {
+		chunks = nodes;
+		INFO("reducing chunks to ", chunks, " chunks");
+	}
+	nodesPerChunk = splitNodes(nodes, chunks);
+
+	// compute prefix sum
+	count sum = 0;
+	size_t prev = 0;
+	std::vector<count> nodePrefSum;
+	G.forNodes([&](node u) {
+		size_t adj_size = 0;
+		sum += G.degree(u);
+		nodePrefSum.push_back(sum);
+	});
```

Comment:
AFAICS we do not need the `nodePrefSum` array. We can compute it on-the-fly while writing the file.